### PR TITLE
mapnik: unstable-2022-10-18 -> unstable-2023-11-28

### PR DESCRIPTION
**Pull request by hummeltech**
## Description of changes

* Removed no-longer-needed patch:
  * `include.patch`: Fixed [here](https://github.com/mapnik/mapnik/commit/64a031a19bd8f77eb6f3cc7ee64839e9e7770332)

* Added patch to account for `FONTS_INSTALL_DIR`/`PLUGINS_INSTALL_DIR` containing full paths, causing this issue:
  ```
  Broken paths found in a .pc file! /nix/store/...-mapnik-unstable-2023-11-28/lib/pkgconfig/libmapnik.pc
  The following lines have issues (specifically '//' in paths).
  5:fonts_dir=${prefix}//nix/store/...-mapnik-unstable-2023-11-28/lib/mapnik/fonts
  6:plugins_dir=${prefix}//nix/store/...-mapnik-unstable-2023-11-28/lib/mapnik/input
  ```

* Enabled `Checks`, which requires a patch to allow one test to fail:
  * `datasource-ogr-test-should-fail.patch`

* Updated `python-mapnik` to latest upstream
  * Based on the `proj6` branch (which currently has an open pull request)

This "release" also includes a patch which mod_tile is expecting for Mapnik versions >= 4:
Seen [here](https://github.com/NixOS/nixpkgs/commit/a7fc0b6e7016aa62c8a1e1f038d28d4b33f96e5c)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/mapnik/default.nix [hunk 1]
```diff
@@ -83,8 +78,11 @@ stdenv.mkDerivation rec {
 
   cmakeFlags = [
     # Would require qt otherwise.
-    "-DBUILD_DEMO_VIEWER=OFF"
+    "-DBUILD_DEMO_VIEWER:BOOL=OFF"
+    "-DFONTS_INSTALL_DIR:PATH=lib/mapnik/fonts"
+    "-DPLUGINS_INSTALL_DIR:PATH=lib/mapnik/input"
   ];
+  doCheck = true;
```
Comment (dotlambda) [inline pkgs/development/libraries/mapnik/default.nix:85]:
```suggestion

  doCheck = true;
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please squash.

**Review by dotlambda — CHANGES_REQUESTED** [review]
The first line of commit message should be `mapnik: unstable-2022-10-18 -> unstable-2023-11-28`. You can mention the removed patch in the remainder.

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 277128` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages failed to build:</summary>
  <ul>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>mapnik</li>
    <li>perl536Packages.Tirex</li>
    <li>perl536Packages.Tirex.devdoc</li>
    <li>perl538Packages.Tirex</li>
    <li>perl538Packages.Tirex.devdoc</li>
    <li>python310Packages.python-mapnik</li>
    <li>python310Packages.python-mapnik.dist</li>
    <li>python311Packages.python-mapnik</li>
    <li>python311Packages.python-mapnik.dist</li>
  </ul>
</details>

**Comment by hummeltech** [comment]
Hmm, yes I see that. I am not having any issues building it locally, and the [logs](https://logs.ofborg.org/?attempt_id=441ea9d0-3a5f-4dbd-a679-224d48cb4f13&key=nixos%2Fnixpkgs.277128) are not showing any errors. I have added a commit in order to attempt the build with `enableParallelBuilding` disabled, which is what previous versions of `Mapnik` did in order to build certain targets which required more RAM. And since the `x86_64-darwin` builds (which use `-j1`) are all building successfully, I think this might be a valid path. I cannot think of anything else to try at this time.

**Comment by hummeltech** [comment]
Nevermind, upon closer inspection of the logs, it looks like there is still an issue which requires `cmake-harfbuzz.patch`. I have re-included that patch.

File: pkgs/development/libraries/mapnik/default.nix [hunk 2]
```diff
@@ -83,9 +84,13 @@ stdenv.mkDerivation rec {
 
   cmakeFlags = [
     # Would require qt otherwise.
-    "-DBUILD_DEMO_VIEWER=OFF"
+    "-DBUILD_DEMO_VIEWER:BOOL=OFF"
+    "-DFONTS_INSTALL_DIR:PATH=lib/mapnik/fonts"
+    "-DPLUGINS_INSTALL_DIR:PATH=lib/mapnik/input"
```
Comment (dotlambda) [inline pkgs/development/libraries/mapnik/default.nix:89]: What made you add these two lines? Isn't this the default?

Comment (hummeltech) [reply pkgs/development/libraries/mapnik/default.nix]: I added a patch to address this behavior.

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 277128` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python310Packages.python-mapnik</li>
    <li>python310Packages.python-mapnik.dist</li>
    <li>python311Packages.python-mapnik</li>
    <li>python311Packages.python-mapnik.dist</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>mapnik</li>
    <li>perl536Packages.Tirex</li>
    <li>perl536Packages.Tirex.devdoc</li>
    <li>perl538Packages.Tirex</li>
    <li>perl538Packages.Tirex.devdoc</li>
  </ul>
</details>

**Comment by hummeltech** [comment]
I'm not sure why that was no longer working, I added `python3Packages.boost` to `python-mapnik`'s `propagatedBuildInputs`

**Comment by dotlambda** [comment]
I don't think it has anything to do with adding boost to `propagatedBuildInputs`. I get errors like
```
src/mapnik_enumeration.hpp:71:20: error: invalid ‘static_cast’ from type ‘const mapnik::enumeration<mapnik::gamma_method_enum, mapnik::gamma_method_e_to_string, mapnik::gamma_method_e_from_string, mapnik::gamma_method_e_lookup>’ to type ‘long int’
   71 |                 ,  static_cast<long>( v ));
      |                    ^~~~~~~~~~~~~~~~~~~~~~
```
and
```
src/mapnik_enumeration.hpp:71:20: error: invalid ‘static_cast’ from type ‘const mapnik::enumeration<mapnik::gamma_method_enum, mapnik::gamma_method_e_to_string, mapnik::gamma_method_e_from_string, mapnik::gamma_method_e_lookup>’ to type ‘long int’
   71 |                 ,  static_cast<long>( v ));
      |                    ^~~~~~~~~~~~~~~~~~~~~~
```
I think python-mapnik is simply not yet compatible with recent versions of mapnik.
I suggest you open an upstream issue and mark python-mapnik broken.

**Comment by dotlambda** [comment]
https://github.com/mapnik/python-mapnik/issues/270#issuecomment-1653606505 claims to have a fix but after applying that commit using `fetchpatch` it complains about C++20.

**Comment by hummeltech** [comment]
Alright @dotlambda, I figured out the issue and I now have `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` running without errors on my machine. Let me know if it's working for you.

It looks like support for `Proj >= 6` isn't completely merged into the default branch, so I tried out the `proj6` branch (omitting the latest commits which break the compilation.)
https://github.com/mapnik/python-mapnik/compare/master...proj6

**Review by dotlambda — CHANGES_REQUESTED** [review]
The last commit message should be `python311Packages.python-mapnik: unstable-2020-09-08 -> unstable-2023-02-23`.

**Comment by dotlambda** [comment]
@hummeltech Do you want to add yourself as maintainer?

**Comment by hummeltech** [comment]
Sure, that's fine, shall I add myself onto the list?

**Comment by dotlambda** [comment]
> Sure, that's fine, shall I add myself onto the list?

Yes, you'll first have to add yourself to `maintainers/maintainer-list.nix` in one commit and then to mapnik's `meta.maintainers` in a separate one.

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 277128` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>mapnik</li>
    <li>perl536Packages.Tirex</li>
    <li>perl536Packages.Tirex.devdoc</li>
    <li>perl538Packages.Tirex</li>
    <li>perl538Packages.Tirex.devdoc</li>
    <li>python310Packages.python-mapnik</li>
    <li>python310Packages.python-mapnik.dist</li>
    <li>python311Packages.python-mapnik</li>
    <li>python311Packages.python-mapnik.dist</li>
  </ul>
</details>

**Review by dotlambda — APPROVED** [review]
Darwin fails but I don't have access to the darwin builder atm so I won't be able to fix it. That shouldn't stop us from updating though.

**Pull request merged**